pkg/apis/extensions/v1alpha1/register.go

```diff
@@ -44,12 +46,33 @@ var (
 	AddToScheme = SchemeBuilder.AddToScheme
 )
 
+// ObjectWebhook is an interface to use in the webhook.
+type ObjectWebhook interface {
+	runtime.Object
+	apis.Validatable
+	apis.Defaultable
+	kmeta.OwnerRefable
+}
```

Comment:
_Here and in the other `*_register.go` files below:_
I don't think that we need to duplicate interfaces just to pass some additional parameters in one module. If I really wanted to re-use object lists, I'd define a common structure somewhere in `pkg/apis/common/v1alpha1/types.go`:

```go
// GroupObject holds the API group object types.
//
// +k8s:deepcopy-gen=false
type GroupObject struct {
	Single, List runtime.Object
}
```

And then:

```diff
diff --git a/pkg/apis/extensions/v1alpha1/register.go b/pkg/apis/extensions/v1alpha1/register.go
index e65a8af2..ea2c1e53 100644
--- a/pkg/apis/extensions/v1alpha1/register.go
+++ b/pkg/apis/extensions/v1alpha1/register.go
@@ -17,9 +17,8 @@ limitations under the License.
 package v1alpha1

 import (
+       "github.com/triggermesh/triggermesh/pkg/apis/common/v1alpha1"
        "github.com/triggermesh/triggermesh/pkg/apis/extensions"
-       "knative.dev/pkg/apis"
-       "knative.dev/pkg/kmeta"

        metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
        "k8s.io/apimachinery/pkg/runtime"
@@ -46,33 +45,16 @@ var (
        AddToScheme = SchemeBuilder.AddToScheme
 )

-// ObjectWebhook is an interface to use in the webhook.
-type ObjectWebhook interface {
-       runtime.Object
-       apis.Validatable
-       apis.Defaultable
-       kmeta.OwnerRefable
-}
-
-// Objects holds instances of ObjectWebhook and runtime.Objects List.
-// +k8s:deepcopy-gen=false
-type Objects struct {
-       Single ObjectWebhook
-       List   runtime.Object
-}
-
 // AllTypes is a list of all the types defined in this package.
-var AllTypes = []Objects{
-       {&Function{}, &FunctionList{}},
+var AllTypes = []v1alpha1.GroupObject{
+       {Single: &Function{}, List: &FunctionList{}},
 }

 // Adds the list of known types to Scheme.
 func addKnownTypes(scheme *runtime.Scheme) error {
-       allTypes := make([]runtime.Object, 0, len(AllTypes)*2)
        for _, t := range AllTypes {
-               allTypes = append(allTypes, t.Single, t.List)
+               scheme.AddKnownTypes(SchemeGroupVersion, t.Single, t.List)
        }
-       scheme.AddKnownTypes(SchemeGroupVersion, allTypes...)
        metav1.AddToGroupVersion(scheme, SchemeGroupVersion)
        return nil
 }
```

Reply:
@tzununbekov The line that you comment here: `{Single: &Function{}, List: &FunctionList{}},` doesn't seems to be replicable for all components, maybe I'm wrong but I don't think we have these common structs for all groups  like sources, targets, flow and routing. We have a struct for each component, so I guess it'll complicate more the code.

Reply:
> maybe I'm wrong but I don't think we have these common structs for all groups like sources, targets, flow and routing

Which common structs do you mean?

Reply:
Maybe I'm not understanding the solution you propose but I mean for extensions there is only Function and FunctionList but for example for sources we have x and xList per source, that seems like a lot to add, that also happens for others like targets, flow, routing.

Reply:
I did exactly what you proposed with the only difference - I didn't create a new interface for each group but created a common `GroupObject` structure in `pkg/apis/common/v1alpha1/types.go` (as it is mentioned in the first comment) so that all resources can be registered as a Function example in this diff:

```go
var AllTypes = []v1alpha1.GroupObject{
  {Single: &Function{}, List: &FunctionList{}},
}
```
it also works for other resources, eg:

```go
var AllTypes = []v1alpha1.GroupObject{
       {Single: &JQTransformation{}, List: &JQTransformationList{}},
       {Single: &Synchronizer{}, List: &SynchronizerList{}},
       {Single: &Transformation{}, List: &TransformationList{}},
       {Single: &XMLToJSONTransformation{}, List: &XMLToJSONTransformationList{}},
       {Single: &XSLTTransformation{}, List: &XSLTTransformationList{}},
}
```

If it is still confusing, I can push the full version of this update, I have it all stashed locally.

Reply:
Oh I see, thanks